3p-ecosystem/src/main/java/com/google/homesampleapp/screens/home/HomeScreen.kt

```diff
@@ -0,0 +1,731 @@
+import com.google.android.gms.home.matter.commissioning.DeviceInfo
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData.EXTRA_COMMISSIONING_WINDOW_EXPIRATION
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData.EXTRA_DEVICE_NAME
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData.EXTRA_MANUAL_PAIRING_CODE
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData.EXTRA_PRODUCT_ID
+import com.google.android.gms.home.matter.commissioning.SharedDeviceData.EXTRA_VENDOR_ID
+import com.google.android.material.textview.MaterialTextView
+import com.google.homesampleapp.AppViewModel
+import com.google.homesampleapp.Device
+import com.google.homesampleapp.MIN_COMMISSIONING_WINDOW_EXPIRATION_SECONDS
+import com.google.homesampleapp.R
+import com.google.homesampleapp.TaskStatus
+import com.google.homesampleapp.commissioning.AppCommissioningService
+import com.google.homesampleapp.getDeviceTypeIconId
+import com.google.homesampleapp.isMultiAdminCommissioning
+import com.google.homesampleapp.screens.common.DialogInfo
+import com.google.homesampleapp.screens.common.MsgAlertDialog
+import com.google.homesampleapp.screens.shared.UserPreferencesViewModel
+import com.google.homesampleapp.screens.thread.getActivity
+import com.google.homesampleapp.stateDisplayString
+import com.google.protobuf.Timestamp
+import timber.log.Timber
+
+/**
+ * Home screen for the application.
+ *
+ * The Home screen features four sections:
+ * 1. The list of devices currently commissioned into the app's fabric. When the user clicks on a
+ *    device, app flow moves to the Device screen where one can get additional details on the device
+ *    and perform actions on it. Devices are persisted in the DevicesRepository, a Proto Datastore.
+ *    It's possible to hide the devices that are currently offline via a setting in the Settings
+ *    screen.
+ * 2. Top App Bar. Settings icon to navigate to the Settings screen.
+ * 3. "Add Device" button. Triggers the commissioning of a new device.
+ * 4. Codelab information. When the app is first launched, a Dialog is shown to provide information
+ *    about the app's companion codelab. This can be dismissed via a checkbox and the setting is
+ *    persisted in the UserPreferences proto datastore.
+ *
+ * Note:
+ * - The app currently only supports Matter devices with server attribute "ON/OFF".
+ *
+ * TODO:
+ * - Finding out that a device is offline is not working very well. Much work needed there.
+ */
+@Composable
+internal fun HomeRoute(
+  navController: NavController,
+  innerPadding: PaddingValues,
+  userPreferencesViewModel: UserPreferencesViewModel = hiltViewModel(),
+  appViewModel: AppViewModel = hiltViewModel(),
+  homeViewModel: HomeViewModel = hiltViewModel(),
+) {
+  // FIXME[TJ]: does not change the topAppBar title
+  appViewModel.setAppBarTitle("[HOME]")
```

Comment:
This occurs because `hiltViewModel` will return a `ViewModel` scoped to the current backstack entry, so its creating a new `HomeViewModel` instance for each screen. To get one scoped to the root of the nav graph you will need to pass that entry explicitly as seen here: https://developer.android.com/jetpack/compose/libraries#hilt-navigation

Reply:
I don't quite understand how they make it work in the link you provided.
I tried this:

```
    // Home
    composable("home") { backStackEntry ->
      val appEntry = remember(backStackEntry) {
        navController.getBackStackEntry("App")
      }
      val appViewModel = hiltViewModel<AppViewModel>(appEntry)
      HomeRoute(innerPadding, navigateToDevice, appViewModel)
    }
```

but this does not work.
The ViewModel I want to share with all screens is the AppViewModel. Not clear to me how I made a shared version of it accessible to all screens.

Reply:
But in the end, all I want is for each Screen to have the ability to change the title that is defined in the shared Scaffold. What's the proper way to do that?

Reply:
Found a way to make this work. All good for now.

Reply:
I forgot to reply to this yesterday, glad you found a way to make it work! I was going to suggest using the `home` back stack entry when retrieving it, since home would always be on your back stack you would make reuse the viewModel from this back stack entry on all screens.